rustls/src/conn.rs

```diff
@@ -636,11 +639,13 @@ impl<Data> ConnectionCommon<Data> {
                 continue;
             }
 
-            match (eof, until_handshaked, self.is_handshaking()) {
-                (_, true, false) => return Ok((rdlen, wrlen)),
-                (_, false, _) => return Ok((rdlen, wrlen)),
-                (true, true, true) => return Err(io::Error::from(io::ErrorKind::UnexpectedEof)),
-                (..) => {}
+            let blocked = blocked_write.zip(blocked_read);
+            match (eof, until_handshaked, self.is_handshaking(), blocked) {
+                (_, true, false, _) => return Ok((rdlen, wrlen)),
+                (_, false, _, _) => return Ok((rdlen, wrlen)),
```

Comment:
I think we should check the values of `rdlen`,  `wrlen` and `blocked` here. As in, if `rdlen == wrlen == 0` and `blocked.is_some()` then we should return `Err(WouldBlock)` instead.